internal/wasm/store.go

```diff
@@ -484,28 +505,29 @@ func (s *Store) Instantiate(module *Module, name string) (*PublicModule, error)
 
 // ReleaseModuleInstance deallocates resources if a module with the given name exists.
 func (s *Store) ReleaseModuleInstance(moduleName string) error {
-	// TODO: none of this is goroutine safe
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
 	instance := s.moduleInstances[moduleName]
 	if instance == nil {
 		return nil // already released
 	}
 
-	instance.refCount--
-	if instance.refCount > 0 {
+	instance.mux.Lock()
+	defer instance.mux.Unlock()
+
+	if instance.importedCount > 0 {
 		// This case other modules are importing this module instance and still alive.
-		return fmt.Errorf("%d modules import this and need to be closed first", instance.refCount)
+		return fmt.Errorf("%d modules import this and need to be closed first", instance.importedCount)
 	}
 
 	// TODO: check outstanding calls and wait until they exit.
```

Comment:
this will be resolved in the follow-up and then we could expose Close (or whatever we call) in the public API